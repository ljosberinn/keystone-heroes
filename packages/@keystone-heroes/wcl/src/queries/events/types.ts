export type BaseEvent<T extends Record<string, unknown>> = T & {
  timestamp: number;
  // pin: "0" but irrelevant
};

type EncounterStartEvent = BaseEvent<{
  type: "encounterstart";
  encounterID: number;
  name: string;
  difficulty: number;
  size: number;
  level: number;
  affixes: number[];
}>;

type CombatantInfoEvent = BaseEvent<{
  type: "combatantinfo";
  sourceID: number;
  gear: [];
  auras: {
    source: number;
    ability: number;
    stacks: number;
    icon: string;
    name?: string;
  }[];
  expansion: string;
  faction: number;
  specID: number;
  covenantID: number;
  soulbindID: number;
  strength: number;
  agility: number;
  stamina: number;
  intellect: number;
  dodge: number;
  parry: number;
  block: number;
  armor: number;
  critMelee: number;
  critRanged: number;
  critSpell: number;
  speed: number;
  leech: number;
  hasteMelee: number;
  hasteRanged: number;
  hasteSpell: number;
  avoidance: number;
  mastery: number;
  versatilityDamageDone: number;
  versatilityHealingDone: number;
  versatilityDamageReduction: number;
  talents: { id: number; icon: string }[];
  pvpTalents: { id: number; icon: string }[];
  artifact: { traitID: number; rank: number; spellID: number; icon: string }[];
  heartOfAzeroth: {
    traitID: number;
    rank: number;
    spellID: number;
    icon: string;
  }[];
}>;

export type ApplyDebuffEvent = BaseEvent<{
  type: "applydebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  targetInstance?: number;
  targetMarker?: number;
}>;

export type ApplyDebuffStackEvent = BaseEvent<{
  type: "applydebuffstack";
  stack: number;
  abilityGameID: number;
}>;

export type ApplyBuffEvent = BaseEvent<{
  type: "applybuff";
  sourceID: number;
  sourceInstance?: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
}>;

export type RemoveBuffEvent = BaseEvent<{
  type: "removebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  sourceMarker: number;
  targetMarker: number;
}>;

export type CastEvent = BaseEvent<{
  type: "cast";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  sourceMarker?: number;
}>;

export type DamageEvent = BaseEvent<{
  type: "damage";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  hitType: number;
  amount: number;
  mitigated?: number;
  unmitigatedAmount?: number;
  tick?: boolean;
  sourceMarker?: number;
  absorbed?: number;
  overkill?: number;
  buffs?: string;
}>;

export type BeginCastEvent = BaseEvent<{
  type: "begincast";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
}>;

export type RemoveDebuffEvent = BaseEvent<{
  type: "removedebuff";
  sourceID: number;
  targetID: number;
  targetInstance: number;
  abilityGameID: number;
}>;

type SummonEvent = BaseEvent<{
  type: "summon";
  sourceID: number;
  targetID: number;
  targetInstance: number;
  abilityGameID: number;
}>;

type PhaseStartEvent = BaseEvent<{
  type: "phasestart";
  encounterID: number;
  name: string;
  difficulty: number;
  size: number;
}>;

export type HealEvent = BaseEvent<{
  type: "heal";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  hitType: number;
  amount: number;
  sourceMarker?: number;
  targetMarker?: number;
  tick?: boolean;
  overheal?: number;
  absorbed?: number;
}>;

type EnergizeEvent = BaseEvent<{
  type: "energize";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  resourceChange: number;
  resourceChangeType: number;
  otherResourceChange: number;
  waste: number;
  sourceMarker?: number;
  targetMarker?: number;
}>;

export type ApplyBuffStackEvent = BaseEvent<{
  type: "applybuffstack";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  stack: number;
  sourceMarker?: number;
  targetMarker?: number;
}>;

export type InterruptEvent = BaseEvent<{
  type: "interrupt";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  extraAbilityGameID: number;
  sourceMarker?: number;
}>;

export type AbsorbEvent = BaseEvent<{
  type: "absorbed";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  attackerID: number;
  amount: number;
  extraAbilityGameID: number;
  sourceMarker: number;
  targetMarker: number;
}>;

export type RefreshDebuffEvent = BaseEvent<{
  type: "refreshdebuff";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
}>;

type DispelEvent = BaseEvent<{
  type: "dispel";
  sourceID: number;
  targetID: number;
  abilityGameID: number;
  extraAbilityGameID: number;
  isBuff: boolean;
  sourceMarker: number;
  targetMarker: number;
}>;

export type DeathEvent = BaseEvent<{
  type: "death";
  sourceID: number;
  targetID: number;
  targetInstance?: number;
  abilityGameID: number;
  // only present if NPC kills player
  killerID?: number;
  killerInstance?: number;
  killingAbilityGameID?: number;
  targetMarker?: number;
}>;

export type AnyEvent =
  | DeathEvent
  | DispelEvent
  | RefreshDebuffEvent
  | AbsorbEvent
  | InterruptEvent
  | ApplyBuffStackEvent
  | EnergizeEvent
  | HealEvent
  | PhaseStartEvent
  | SummonEvent
  | RemoveDebuffEvent
  | BeginCastEvent
  | DamageEvent
  | CastEvent
  | RemoveBuffEvent
  | ApplyBuffEvent
  | ApplyDebuffStackEvent
  | ApplyDebuffEvent
  | CombatantInfoEvent
  | EncounterStartEvent;

export type AllTrackedEventTypes = (
  | CastEvent
  | DeathEvent
  | AbsorbEvent
  | DamageEvent
  | HealEvent
  | InterruptEvent
  | BeginCastEvent
  | ApplyBuffEvent
  | ApplyDebuffEvent
  | ApplyDebuffStackEvent
  | RemoveBuffEvent
  | ApplyBuffStackEvent
)[];
