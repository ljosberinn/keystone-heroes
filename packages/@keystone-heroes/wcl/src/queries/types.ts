import type { PlayableClass, Role, SpecName } from "@prisma/client";

export type Table = {
  totalTime: number;
  itemLevel: number;
  logVersion: number;
  gameVersion: number;
  composition: Composition[];
  damageDone: DamageDone[];
  healingDone: HealingDone[];
  damageTaken: DamageTaken[];
  deathEvents: DeathEvent[];
  playerDetails: PlayerDetails;
};

type Composition = {
  name: string;
  id: number;
  guid: number;
  type: PlayableClass;
  specs: SpecMeta[];
};

type SpecMeta = {
  spec: SpecName;
  role: Role;
};

export type DamageDone = Pick<Composition, "name" | "id" | "guid"> & {
  total: number;
  icon: string;
  type: Composition["type"] | "Pet" | "NPC";
};

export type HealingDone = DamageDone;

export type DamageTaken = Pick<DamageDone, "name" | "guid" | "total"> & {
  abilityIcon: string;
  type: number;
};

type DeathEvent = Pick<DamageDone, "name" | "id" | "guid" | "type" | "icon"> & {
  deathTime: number;
  ability: Omit<DamageTaken, "total">;
};

export type PlayerDetails = {
  tanks: InDepthCharacterInformation[];
  dps: InDepthCharacterInformation[];
  healers: InDepthCharacterInformation[];
};

export type InDepthCharacterInformation = {
  name: string;
  id: number;
  guid: number;
  type: PlayableClass;
  server: string;
  icon: string;
  specs: SpecName[];
  minItemLevel: number;
  maxItemLevel: number;
  combatantInfo: CombatantInfo;
};

type CombatantInfo = {
  stats: Stats;
  talents: Talent[];
  gear: (Item | LegendaryItem)[];
  /**
   * @deprecated https://canary.discord.com/channels/180033360939319296/681904912090529801/863865520486088704
   */
  artifact?: CustomPowerSet[];
  /**
   * @deprecated https://canary.discord.com/channels/180033360939319296/681904912090529801/863865520486088704
   */
  heartOfAzeroth?: SecondaryCustomPowerSet[];
  customPowerSet: CustomPowerSet[];
  secondaryCustomPowerSet: SecondaryCustomPowerSet[];
  tertiaryCustomPowerSet: unknown[];
  specIDs: number[];
  factionID: number;
  covenantID?: number;
  soulbindID?: number;
};

type Stats = {
  Speed: Stat;
  Dodge: Stat;
  Mastery: Stat;
  Stamina: Stat;
  Haste: Stat;
  Leech: Stat;
  Armor: Stat;
  Agility: Stat;
  Crit: Stat;
  "Item Level": Stat;
  Parry: Stat;
  Avoidance: Stat;
  Versatility: Stat;
  Intellect: Stat;
};

type Stat = {
  min: number;
  max: number;
};

export type Talent = {
  name: string;
  guid: number;
  type: number;
  abilityIcon: string;
};

type CustomPowerSet = {
  name: string;
  guid: number;
  type: number;
  abilityIcon: string;
  total: number;
};

export enum ItemQuality {
  POOR = 1,
  COMMON = 2,
  RARE = 3,
  SUPERIOR = 4,
  LEGENDARY = 5,
}

export type Item = {
  id: number;
  slot: number;
  quality: ItemQuality;
  icon: string;
  name?: string;
  itemLevel: number;
  bonusIDs?: number[];
  gems?: Gem[];
  permanentEnchant?: number;
  permanentEnchantName?: string;
  onUseEnchant?: number;
  onUseEnchantName?: string;
  effectID?: number;
  effectIcon?: string;
  effectName?: string;
  temporaryEnchant?: number;
  temporaryEnchantName?: string;
};

export type LegendaryItem = Item & {
  quality: ItemQuality.LEGENDARY;
  effectID: number;
  effectIcon: string;
  effectName: string;
};

type Gem = {
  id: number;
  itemLevel: number;
  icon: string;
};

type SecondaryCustomPowerSet = {
  name: string;
  guid: number;
  type: number;
  abilityIcon: string;
  total: number;
};

export type Conduit = SecondaryCustomPowerSet;
export type CovenantTrait = CustomPowerSet;
