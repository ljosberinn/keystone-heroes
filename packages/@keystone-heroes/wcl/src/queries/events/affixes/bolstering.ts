import { Affixes } from "@keystone-heroes/db/types";

import type { AllTrackedEventTypes, ApplyBuffEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const BOLSTERING = 209_859;

/**
 * @see https://www.warcraftlogs.com/reports/mBf1t94GZ2hRWprv#fight=7&type=auras&view=events&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22applybuff%22%20and%20ability.id%20%3D%20209859%20and%20target.type%20%3D%20%22NPC%22&hostility=1
 *
 * ```gql
 * {
 *   reportData {
 *     report(code: "mBf1t94GZ2hRWprv") {
 *       fights(fightIDs: [7]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 5996572, endTime: 7458759, limit: 2000, filterExpression: "type = \"applybuff\" and ability.id = 209859 and target.type = \"NPC\"") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "applybuff" and ability.id = ${BOLSTERING} and target.type = "npc"`,
];

const isBolsteringEvent = createIsSpecificEvent<ApplyBuffEvent>({
  type: "applybuff",
  abilityGameID: BOLSTERING,
});

type StackDataset = {
  targetID: number;
  targetInstance: number | null;
  stacks: number;
};

const getHighestBolsteringStack = (
  allEvents: ApplyBuffEvent[]
): ApplyBuffEvent[] => {
  const stacksPerNPC = allEvents.reduce<StackDataset[]>(
    (acc, { targetID, targetInstance = null }) => {
      const existingIndex = acc.findIndex(
        (dataset) =>
          dataset.targetInstance === targetInstance &&
          targetID === dataset.targetID
      );

      if (existingIndex > -1) {
        return acc.map((dataset, index) =>
          index === existingIndex
            ? { ...dataset, stacks: dataset.stacks + 1 }
            : dataset
        );
      }

      return [
        ...acc,
        {
          targetID,
          targetInstance,
          stacks: 1,
        },
      ];
    },
    []
  );

  const highestStack = stacksPerNPC.reduce((acc, dataset) =>
    acc?.stacks >= dataset.stacks ? acc : dataset
  );

  const NPCsWithSameFinalStackAmount = stacksPerNPC.filter(
    (dataset) => dataset.stacks === highestStack.stacks
  );

  return allEvents.reduce<(ApplyBuffEvent & { stacks: number })[]>(
    (acc, event) => {
      const dataset = NPCsWithSameFinalStackAmount.find(
        (npc) =>
          npc.targetID === event.targetID &&
          npc.targetInstance === event.targetInstance
      );

      if (!dataset) {
        return acc;
      }

      const previousMatch = acc.find(
        (oldEvent) =>
          oldEvent.targetID === event.targetID &&
          oldEvent.targetInstance === event.targetInstance
      );

      const enhancedEvent = { ...event, stacks: dataset.stacks };

      if (!previousMatch) {
        return [...acc, enhancedEvent];
      }

      // implicitly means it must be a higher stack amount as targetID and
      // targetInstance match
      const isNewer = event.timestamp > previousMatch.timestamp;

      if (isNewer) {
        return [
          ...acc.filter((oldEvent) => oldEvent !== previousMatch),
          enhancedEvent,
        ];
      }

      return acc;
    },
    []
  );
};

export const getBolsteringEvents = (
  allEvents: AllTrackedEventTypes,
  affixSet: Set<Affixes>
): ApplyBuffEvent[] => {
  if (!affixSet.has(Affixes.Bolstering)) {
    return [];
  }

  return getHighestBolsteringStack(allEvents.filter(isBolsteringEvent));
};
