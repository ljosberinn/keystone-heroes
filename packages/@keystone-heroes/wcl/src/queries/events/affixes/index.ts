import { Affixes } from "@keystone-heroes/db/types";

import type { EventParams } from "..";
import type {
  AbsorbEvent,
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyDebuffStackEvent,
  DamageEvent,
  HealEvent,
  InterruptEvent,
} from "../types";
import {
  filterExpression as bolsteringFilterExpression,
  getBolsteringEvents,
} from "./bolstering";
import {
  filterExpression as burstingFilterExpression,
  getBurstingEvents,
} from "./bursting";
import {
  filterExpression as explosiveFilterExpression,
  getExplosiveEvents,
} from "./explosive";
import {
  filterExpression as grievousFilterExpression,
  getGrievousEvents,
} from "./grievous";
import {
  filterExpression as necroticFilterExpression,
  getNecroticEvents,
} from "./necrotic";
import {
  filterExpression as quakingFilterExpression,
  getQuakingEvents,
} from "./quaking";
import {
  filterExpression as sanguineFilterExpression,
  getSanguineEvents,
} from "./sanguine";
import {
  filterExpression as spitefulFilterExpression,
  getSpitefulEvents,
} from "./spiteful";
import {
  filterExpression as stormingFilterExpression,
  getStormingEvents,
} from "./storming";
import {
  filterExpression as tormentedFilterExpression,
  getTormentedEvents,
} from "./tormented";
import {
  filterExpression as volcanicFilterExpression,
  getVolcanicEvents,
} from "./volcanic";

type AffixWithEvents = Exclude<
  Affixes,
  | "Awakened"
  | "Beguiling"
  | "Tyrannical"
  | "Fortified"
  | "Prideful"
  | "Raging"
  | "Infested"
  | "Inspiring"
  | "Reaping"
  | "Skittish"
>;

const affixExpressionMap: Record<AffixWithEvents, string[]> = {
  [Affixes.Sanguine]: sanguineFilterExpression,
  [Affixes.Explosive]: explosiveFilterExpression,
  [Affixes.Grievous]: grievousFilterExpression,
  [Affixes.Necrotic]: necroticFilterExpression,
  [Affixes.Volcanic]: volcanicFilterExpression,
  [Affixes.Bursting]: burstingFilterExpression,
  [Affixes.Spiteful]: spitefulFilterExpression,
  [Affixes.Quaking]: quakingFilterExpression,
  [Affixes.Storming]: stormingFilterExpression,
  [Affixes.Bolstering]: bolsteringFilterExpression,
  [Affixes.Tormented]: tormentedFilterExpression,
};

export const getAffixExpression = (affixes: Affixes[]): string[] => {
  return affixes
    .filter((affix): affix is AffixWithEvents => affix in affixExpressionMap)
    .flatMap((affix) => affixExpressionMap[affix]);
};

export const filterAffixEvents = (
  allEvents: AllTrackedEventTypes[],
  affixes: EventParams["affixes"]
): (
  | AbsorbEvent
  | InterruptEvent
  | HealEvent
  | DamageEvent
  | ApplyBuffEvent
  | ApplyDebuffStackEvent
)[] => {
  const affixSet = new Set(affixes);

  // seasonal
  const tormentedEvents = getTormentedEvents(allEvents, affixSet);

  // common affixes
  const storming = getStormingEvents(allEvents, affixSet);
  const spiteful = getSpitefulEvents(allEvents, affixSet);
  const sanguine = getSanguineEvents(allEvents, affixSet);
  const volcanic = getVolcanicEvents(allEvents, affixSet);
  const quaking = getQuakingEvents(allEvents, affixSet);
  const bolstering = getBolsteringEvents(allEvents, affixSet);
  const bursting = getBurstingEvents(allEvents, affixSet);
  const explosive = getExplosiveEvents(allEvents, affixSet);
  const grievous = getGrievousEvents(allEvents, affixSet);
  const necrotic = getNecroticEvents(allEvents, affixSet);

  return [
    ...tormentedEvents,
    ...storming,
    ...spiteful,
    ...sanguine,
    ...volcanic,
    ...quaking,
    ...bolstering,
    ...bursting,
    ...explosive,
    ...grievous,
    ...necrotic,
  ];
};
