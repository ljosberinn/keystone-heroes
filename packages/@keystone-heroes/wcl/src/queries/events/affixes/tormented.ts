import { Affixes } from "@keystone-heroes/db/types";

import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
} from "../types";

export const tormentedLieutenants = [
  {
    id: 179_891,
    name: "Soggodon the Breaker",
    icon: "inv_icon_wingbroken04a",
  },
  {
    id: 179_446,
    name: "Incinerator Arkolath",
    icon: "ability_warlock_fireandbrimstone",
  },
  {
    id: 179_892,
    name: "Oros Coldheart",
    icon: "spell_shadow_soulleech_2",
  },
  {
    id: 179_890,
    name: "Executioner Varruth",
    icon: "spell_misc_emotionafraid",
  },
];

export const tormentedLieutenantIDSet = new Set(
  tormentedLieutenants.map((dataset) => dataset.id)
);

const SOGGODON_THE_BREAKER = {
  STONE_WARD: 357_524,
  DRIPPING_FANG: 356_828,
  SELF_EMBALMING_KIT: 357_556,
  TINY_DANCING_SHOES: 356_827,
  BROKEN_MIRROR: 357_778,
};

const INCINERATOR_ARKOLATH = {
  THE_FIFTH_SKULL: 357_839,
  CHAMPIONS_BRAND: 357_575,
  RAGING_BATTLE_AXE: 357_864,
  SIGNET_OF_BOLSTERING: 357_848,
  PEDESTAL_OF_UTTER_HUBRIS: 357_889,
  CRUMBLING_BULWARK: 357_897,
  BOTTLE_OF_SANGUINE_ICHOR: 357_900,
};

const OROS_COLDHEART = {
  REGENERATIVE_FUNGUS: 357_814,
  VIAL_OF_DESPERATION: 357_825,
  HANDBOOK_OF_UNCIVIL_ETIQUETTE: 357_834,
  SATCHEL_OF_THE_HUNT: 357_815,
  HUNTMANS_HORN: 357_817,
  PENDANT_OF_THE_MARTYR: 357_820,
  PORTABLE_FEEDING_TROUGH: 357_842,
  GAVEL_OF_JUDGEMENT: 357_829,
};

const EXECUTIONER_VARRUTH = {
  CHAMPIONS_BRAND: INCINERATOR_ARKOLATH.CHAMPIONS_BRAND,
  DAGGER_OF_NECROTIC_WOUNDING: 357_609,
  VOLCANIC_PLUMAGE: 357_706,
  STYGIAN_KINGS_BARBS: 357_863,
  SIEGEBREAKERS_STAND: 357_604,
  OVERFLOWING_CHALICE: 357_747,
  STABILIZING_DIAMONT_ALEMBIC: 357_847,
};

export const TORMENTED = {
  ...SOGGODON_THE_BREAKER,
  ...INCINERATOR_ARKOLATH,
  ...OROS_COLDHEART,
  ...EXECUTIONER_VARRUTH,
};

/**
 * @TODO remove once WCL provides this itself
 */
export const tormentedSpells = [
  {
    id: TORMENTED.STONE_WARD,
    name: "The Stone Ward",
    icon: "ability_mage_shattershield",
  },
  {
    id: TORMENTED.DRIPPING_FANG,
    name: "Dripping Fang",
    icon: "inv_misc_monsterfang_01",
  },
  {
    id: TORMENTED.TINY_DANCING_SHOES,
    name: "Tiny Dancing Shoes",
    icon: "inv_boots_pvppriest_e_01",
  },
  {
    id: TORMENTED.BROKEN_MIRROR,
    name: "Broken Mirror",
    icon: "trade_archaeology_highbornesoulmirror",
  },
  {
    id: TORMENTED.SELF_EMBALMING_KIT,
    name: "Self-Embalming Kit",
    icon: "inv_mummypet",
  },
  {
    id: TORMENTED.RAGING_BATTLE_AXE,
    name: "Raging Battle-Axe",
    icon: "ability_hunter_swiftstrike",
  },
  {
    id: TORMENTED.SIGNET_OF_BOLSTERING,
    name: "Signet of Bolstering",
    icon: "ability_socererking_arcanefortification",
  },
  {
    id: TORMENTED.CRUMBLING_BULWARK,
    name: "Crumbling Bulwark",
    icon: "trade_archaeology_stoneshield",
  },
  {
    id: TORMENTED.THE_FIFTH_SKULL,
    name: "The Fifth Skull",
    icon: "inv_misc_bone_skull_02",
  },
  {
    id: TORMENTED.BOTTLE_OF_SANGUINE_ICHOR,
    name: "Bottle of Sanguine Ichor",
    icon: "inv_potion_27",
  },
  {
    id: TORMENTED.PEDESTAL_OF_UTTER_HUBRIS,
    name: "Pedestal of Utter Hubris",
    icon: "ability_paladin_beaconoflight",
  },
  {
    id: TORMENTED.VIAL_OF_DESPERATION,
    name: "Vial of Desperation",
    icon: "trade_archaeology_crackedcrystalvial",
  },
  {
    id: TORMENTED.HANDBOOK_OF_UNCIVIL_ETIQUETTE,
    name: "Handbook of Uncivil Etiquette",
    icon: "inv_misc_profession_book_cooking",
  },
  {
    id: TORMENTED.HUNTMANS_HORN,
    name: "Huntman's Horn",
    icon: "inv_misc_horn_01",
  },
  {
    id: TORMENTED.PENDANT_OF_THE_MARTYR,
    name: "Pendant of the Martyr",
    icon: "inv_jewelry_necklace_75",
  },
  {
    id: TORMENTED.PORTABLE_FEEDING_TROUGH,
    name: "Portable Feeding Trough",
    icon: "inv_crate_07",
  },
  {
    id: TORMENTED.REGENERATIVE_FUNGUS,
    name: "Regenerative Fungus",
    icon: "inv_misc_starspecklemushroom",
  },
  {
    id: TORMENTED.SATCHEL_OF_THE_HUNT,
    name: "Satchel of the Hunt",
    icon: "inv_misc_coinbag11",
  },
  {
    id: TORMENTED.GAVEL_OF_JUDGEMENT,
    name: "Gavel of Judgement",
    icon: "inv_hammer_17",
  },
  {
    id: TORMENTED.DAGGER_OF_NECROTIC_WOUNDING,
    name: "Dagger of Necrotic Wounding",
    icon: "inv_glaive_1h_maldraxxusquest_b_01",
  },
  {
    id: TORMENTED.VOLCANIC_PLUMAGE,
    name: "Volcanig Plumage",
    icon: "artifactability_firemage_phoenixbolt",
  },
  {
    id: TORMENTED.STYGIAN_KINGS_BARBS,
    name: "The Stygian King's Barbs",
    icon: "inv_misc_herb_goldthorn_bramble",
  },
  {
    id: TORMENTED.SIEGEBREAKERS_STAND,
    name: "Siegebreaker's Stand",
    icon: "ability_warrior_unrelentingassault",
  },
  {
    id: TORMENTED.OVERFLOWING_CHALICE,
    name: "Overflowing Chalice",
    icon: "inv_drink_22",
  },
  {
    id: TORMENTED.STABILIZING_DIAMONT_ALEMBIC,
    name: "Stabilizing Diamond Alembic",
    icon: "inv_trinket_80_alchemy01",
  },
  {
    id: TORMENTED.CHAMPIONS_BRAND,
    name: "Champion's Brand",
    icon: "spell_warrior_sharpenblade",
  },
];

/**
 * @see https://www.warcraftlogs.com/reports/J3WKacdjpntmLT7C/#fight=3&type=damage-done&view=events&pins=2%24Off%24%23244F4B%24expression%24(type%20%3D%20%22damage%22%20and%20ability.id%20in%20(357865,%20357841,%20357708,%20357901,%20357525,%20356925,%20356923,%20355806,%20358784,%20358970,%20356667,%20356414,%20358894,%20355709,%20355737,%20358967))%20or%20(type%20%3D%20%22heal%22%20and%20ability.id%20%3D%20357901)
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "DrkYtWa2ZPXQ8BjT") {
 *       fights(fightIDs: [2]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 3895428, endTime: 6047081, filterExpression: "type in (\"applybuff\", \"applybuffstack\") and ability.id in (357524, 356828, 357556, 356827, 357778, 357839, 357575, 357864, 357848, 357889, 357897, 357900, 357814, 357825, 357834, 357815, 357817, 357820, 357842, 357829, 357609, 357706, 357863, 357604, 357747, 357847)") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type in ("applybuff", "applybuffstack") and ability.id in (${[
    Object.values(TORMENTED),
  ].join(", ")})`,
];

export const tormentedAbilityGameIDSet = new Set(Object.values(TORMENTED));

export const getTormentedEvents = (
  allEvents: AllTrackedEventTypes[],
  affixSet: Set<Affixes>
): (ApplyBuffEvent | ApplyBuffStackEvent)[] => {
  if (!affixSet.has(Affixes.Tormented)) {
    return [];
  }

  return allEvents.reduce<(ApplyBuffEvent | ApplyBuffStackEvent)[]>(
    (acc, event) => {
      if (event.type !== "applybuff" && event.type !== "applybuffstack") {
        return acc;
      }

      if (!tormentedAbilityGameIDSet.has(event.abilityGameID)) {
        return acc;
      }

      // on death, powers get reapplied and we don't care about those
      if (
        acc.some(
          (prevEvent) =>
            event.targetID === prevEvent.targetID &&
            event.abilityGameID === prevEvent.abilityGameID &&
            event.type === prevEvent.type
        )
      ) {
        return acc;
      }

      return [...acc, event];
    },
    []
  );
};
