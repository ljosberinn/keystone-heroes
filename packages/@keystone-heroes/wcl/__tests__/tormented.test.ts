import { Affixes } from "@keystone-heroes/db/types";
import { getTormentedEvents } from "@keystone-heroes/wcl/queries/events/affixes/tormented";

import allEvents from "./fixtures/allEvents.json";
import earlyPTRTormented from "./fixtures/earlyPTRTormented.json";

describe("getTormentedEvents", () => {
  test("works", () => {
    expect(
      getTormentedEvents(
        [...allEvents, ...earlyPTRTormented],
        new Set([Affixes.Tormented])
      )
    ).toMatchSnapshot();
  });

  test("does nothing if affix is absent", () => {
    expect(
      getTormentedEvents([...allEvents, ...earlyPTRTormented], new Set())
    ).toHaveLength(0);
  });
});
