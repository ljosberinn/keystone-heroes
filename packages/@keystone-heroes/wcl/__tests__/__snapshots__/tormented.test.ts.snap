// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTormentedEvents works 1`] = `
Array [
  Object {
    "abilityGameID": 357865,
    "amount": 271095,
    "hitType": 1,
    "pin": "0",
    "sourceID": 3,
    "sourceMarker": 7,
    "targetID": 61,
    "targetInstance": 5,
    "timestamp": 3270126,
    "type": "damage",
    "unmitigatedAmount": 2780,
  },
  Object {
    "abilityGameID": 357708,
    "amount": 792735,
    "hitType": 1,
    "pin": "0",
    "sourceID": 5,
    "targetID": 61,
    "targetInstance": 5,
    "timestamp": 3277294,
    "type": "damage",
    "unmitigatedAmount": 9881,
  },
  Object {
    "abilityGameID": 357708,
    "amount": 302083,
    "hitType": 2,
    "pin": "0",
    "sourceID": 1,
    "targetID": 61,
    "targetInstance": 5,
    "timestamp": 3288394,
    "type": "damage",
    "unmitigatedAmount": 8748,
  },
  Object {
    "abilityGameID": 357708,
    "amount": 600163,
    "hitType": 1,
    "pin": "0",
    "sourceID": 2,
    "targetID": 61,
    "targetInstance": 5,
    "timestamp": 3289276,
    "type": "damage",
    "unmitigatedAmount": 10108,
  },
  Object {
    "abilityGameID": 356414,
    "absorbed": 27446,
    "amount": 0,
    "hitType": 3,
    "mitigated": 3362,
    "pin": "0",
    "sourceID": 22,
    "targetID": 5,
    "timestamp": 3483009,
    "type": "damage",
    "unmitigatedAmount": 30808,
  },
  Object {
    "abilityGameID": 356414,
    "amount": 16768,
    "hitType": 1,
    "mitigated": 14040,
    "pin": "0",
    "sourceID": 22,
    "targetID": 3,
    "targetMarker": 7,
    "timestamp": 3504055,
    "type": "damage",
    "unmitigatedAmount": 30808,
  },
  Object {
    "abilityGameID": 356414,
    "amount": 1,
    "hitType": 1,
    "mitigated": 3769,
    "overkill": 27038,
    "pin": "0",
    "sourceID": 22,
    "targetID": 2,
    "timestamp": 3514978,
    "type": "damage",
    "unmitigatedAmount": 30808,
  },
  Object {
    "abilityGameID": 356667,
    "absorbed": 665,
    "amount": 2689,
    "hitType": 1,
    "mitigated": 2191,
    "pin": "0",
    "sourceID": 22,
    "targetID": 3,
    "targetMarker": 7,
    "tick": true,
    "timestamp": 3479151,
    "type": "damage",
    "unmitigatedAmount": 5545,
  },
  Object {
    "abilityGameID": 356667,
    "absorbed": 4256,
    "amount": 29401,
    "hitType": 1,
    "mitigated": 828,
    "pin": "0",
    "sourceID": 22,
    "targetID": 5,
    "tick": true,
    "timestamp": 3479170,
    "type": "damage",
    "unmitigatedAmount": 9242,
  },
  Object {
    "abilityGameID": 356667,
    "absorbed": 4378,
    "amount": -29235,
    "hitType": 1,
    "mitigated": 952,
    "pin": "0",
    "sourceID": 22,
    "targetID": 2,
    "tick": true,
    "timestamp": 3479394,
    "type": "damage",
    "unmitigatedAmount": 9242,
  },
  Object {
    "abilityGameID": 356667,
    "amount": 23593,
    "hitType": 1,
    "mitigated": 1377,
    "pin": "0",
    "sourceID": 22,
    "targetID": 60,
    "tick": true,
    "timestamp": 3497512,
    "type": "damage",
    "unmitigatedAmount": 9242,
  },
  Object {
    "abilityGameID": 356667,
    "absorbed": 4489,
    "amount": 38628,
    "hitType": 1,
    "mitigated": 202,
    "pin": "0",
    "sourceID": 22,
    "targetID": 1,
    "tick": true,
    "timestamp": 3497587,
    "type": "damage",
    "unmitigatedAmount": 9242,
  },
  Object {
    "abilityGameID": 356923,
    "absorbed": 9576,
    "amount": 104043,
    "hitType": 1,
    "mitigated": 20544,
    "pin": "0",
    "sourceID": 24,
    "targetID": 3,
    "targetMarker": 7,
    "timestamp": 3187640,
    "type": "damage",
    "unmitigatedAmount": 42359,
  },
  Object {
    "abilityGameID": 356925,
    "absorbed": 9872,
    "amount": 37572,
    "hitType": 1,
    "mitigated": 2706,
    "pin": "0",
    "sourceID": 24,
    "targetID": 1,
    "timestamp": 3179758,
    "type": "damage",
    "unmitigatedAmount": 18676,
  },
  Object {
    "abilityGameID": 356925,
    "absorbed": 428,
    "amount": 86596,
    "hitType": 1,
    "mitigated": 6681,
    "pin": "0",
    "sourceID": 24,
    "targetID": 60,
    "timestamp": 3197963,
    "type": "damage",
    "unmitigatedAmount": 19254,
  },
]
`;
